### PR TITLE
feat: add override_action to aws_networkfirewall_firewall_policy

### LINKED ISSUE #25032: Add support for stateful_rule_group_reference override in AWS managed rules
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
As a follow up to https://github.com/hashicorp/terraform-provider-aws/pull/22355, kindly add support for Override in the `stateful_rule_group_reference` for aws_networkfirewall_firewall_policy resource.

**Currently supported**
```
resource "aws_networkfirewall_firewall_policy" "example" {
  name = "example"

  firewall_policy {
    stateless_default_actions          = ["aws:pass"]
    stateless_fragment_default_actions = ["aws:drop"]
    stateless_rule_group_reference {
      priority     = 1
      resource_arn = aws_networkfirewall_rule_group.example.arn

    stateful_rule_group_reference {
      priority     = 2
      resource_arn = aws_networkfirewall_rule_group.example.arn
    }
  }`
```

 **feature request**
```
resource "aws_networkfirewall_firewall_policy" "example" {
  name = "example"

  firewall_policy {
    stateless_default_actions          = ["aws:pass"]
    stateless_fragment_default_actions = ["aws:drop"]
    stateless_rule_group_reference {
      priority     = 1
      overrride {
             action = ["DROP_TO_ALERT"]
      }
      resource_arn = aws_networkfirewall_rule_group.example.arn

    stateful_rule_group_reference {
      priority     = 2
      overrride {
             action = ["DROP_TO_ALERT"]
      }
      resource_arn = aws_networkfirewall_rule_group.example.arn
    }
  }`
```

### New or Affected Resource(s)

* aws_XXXXX

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.
https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/network-firewall.html
        'StatefulRuleGroupReferences': [
            {
                'ResourceArn': 'string',
                'Priority': 123,
                'Override': {
                    'Action': 'DROP_TO_ALERT'
                }
            },
        ],
```

### References

* #0000

### PR DESCRIPTION
**Pull request by teddylear**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

## Description
Add `overide_action` to `aws_networkfirewall_firewall_policy`

## Output from acceptance testing:
```zsh
terraform-provider-aws on  f-aws_networkfirewall_firewall_policy-add-stateless-rule-group-ref-override via 🐹 v1.17.3 took 6m 26s
[I] ➜  make testacc TESTS=TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference PKG=networkfirewall

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 20 -run='TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupRe
ference'  -timeout 180m
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged (156.55s)
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference (190.90s)
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction (314.68s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall    318.072s
```

```zsh
terraform-provider-aws on  f-aws_networkfirewall_firewall_policy-add-stateless-rule-group-ref-override via 🐹 v1.17.3 took 6m 59s
[I] ➜ make testacc TESTS=TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference PKG=networkfirewall

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 20 -run='TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPr
iorityReference'  -timeout 180m
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference (170.63s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall    173.704s
```

```zsh
terraform-provider-aws on  f-aws_networkfirewall_firewall_policy-add-stateless-rule-group-ref-override [!] via 🐹 v1.17.3 took 7s 
[I] ➜ make testacc TESTS=TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference PKG=networkfirewall

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 20 -run='TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOv
errideActionReference'  -timeout 180m
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference (148.27s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall    151.381s
```
Closes #25032

### CODE REVIEW

**Comment by teddylear** [comment]
@justinretzolk Hey I noticed the the `Semgrep` tests failed here, but the error message is unclear. Can you (or another engineer) provide more insights into the specific error with this check? As for the `terrafmt` check, I can fix that

**Comment by justinretzolk** [comment]
Hey @teddylear  👋 Thank you for your contribution, and for checking in on this! I took a quick look and it seems like the errors you were seeing in the tests have cleared up, so this should be good to go as soon as one of the maintainers is able to take a look at it 🙂

File: internal/service/networkfirewall/firewall_policy.go
Comment (GlennChia) [inline internal/service/networkfirewall/firewall_policy.go:86]: From the [API for CreateFirewallPolicy](https://docs.aws.amazon.com/network-firewall/latest/APIReference/API_CreateFirewallPolicy.html), I noticed `action` is a nested argument within `override`. We usually try to follow the API as close as possible as there could be future API enhancements that add new nested arguments to the parent argument. Keeping to the API documentation, this argument can potentially be the following. We can also add validation for the `action` argument following the documentation [here](https://docs.aws.amazon.com/network-firewall/latest/APIReference/API_StatefulRuleGroupOverride.html).

```suggestion
									"override": {
										Type:     schema.TypeList,
										MaxItems: 1,
										Optional: true,
										Elem: &schema.Resource{
											Schema: map[string]*schema.Schema{
												"action": {
													Type:         schema.TypeString,
													Required:     true,
													ValidateFunc: validation.StringInSlice(networkfirewall.OverrideAction_Values(), false),
												},
											},
										},
									},
```

File: internal/service/networkfirewall/firewall_policy.go
Comment (GlennChia) [inline internal/service/networkfirewall/firewall_policy.go:338]: Since we are using `action` nested in `override`, we will use another `expand` function in order to extract the `action` argument.

File: internal/service/networkfirewall/firewall_policy.go
Comment (GlennChia) [inline internal/service/networkfirewall/firewall_policy.go:453]: Similarly for the read, we will use a `flatten` function to extract `action` from `override`

File: internal/service/networkfirewall/firewall_policy_test.go
Comment (GlennChia) [inline internal/service/networkfirewall/firewall_policy_test.go:335]: Great reference to the valid value documentation for `DROP_TO_ALERT`. As an enhancement, instead of hardcoding this string, we can reference this string using `networkfirewall.OverrideActionDropToAlert`

```suggestion
	override_action := networkfirewall.OverrideActionDropToAlert

File: website/docs/r/networkfirewall_firewall_policy.html.markdown
Comment (GlennChia) [inline website/docs/r/networkfirewall_firewall_policy.html.markdown:106]: We will need to update the documentation to reflect that `action` is a nested argument of `override`

**Review by GlennChia — CHANGES_REQUESTED** [review]
Hi @teddylear, thanks for the PR! This is a really useful addition that will benefit many practitioners. I left some review comments that are mainly to align the resource configuration to the API documentation where I noticed `action` was a nested argument in `override`. We usually follow the API documentation closely so that we can easily add new nested arguments should they be introduced in the future without making breaking changes. Let me know if you have any questions!

File: internal/service/networkfirewall/firewall_policy.go
Comment (teddylear) [reply internal/service/networkfirewall/firewall_policy.go]: Updated.

File: internal/service/networkfirewall/firewall_policy.go
Comment (teddylear) [reply internal/service/networkfirewall/firewall_policy.go]: Updated.

File: internal/service/networkfirewall/firewall_policy_test.go
Comment (teddylear) [reply internal/service/networkfirewall/firewall_policy_test.go]: Updated.

File: website/docs/r/networkfirewall_firewall_policy.html.markdown
Comment (teddylear) [reply website/docs/r/networkfirewall_firewall_policy.html.markdown]: Updated.

**Comment by teddylear** [comment]
@GlennChia Thanks for the review! I updated the PR from your comments above. One question I had was for the test I had to update the attribute I was checking on [this line](https://github.com/teddylear/terraform-provider-aws/blob/f-aws_networkfirewall_firewall_policy-add-stateless-rule-group-ref-override/internal/service/networkfirewall/firewall_policy_test.go#L347). Is there a way I can update the attribute from `firewall_policy.0.stateful_rule_group_reference.0.override.0.action` to `firewall_policy.0.stateful_rule_group_reference.override.action`? I'm assuming there should be only one override block and one action per `stateful_rule_group_reference`, so having an index on it does't make sense here.

File: internal/service/networkfirewall/firewall_policy.go
Comment (GlennChia) [inline internal/service/networkfirewall/firewall_policy.go:91]: I realised upon a second look at the API reference that `action` is optional. [Reference](https://docs.aws.amazon.com/network-firewall/latest/APIReference/API_StatefulRuleGroupOverride.html).

```suggestion
													Optional:     true,

**Review by GlennChia — CHANGES_REQUESTED** [review]
Hi @teddylear, you're right to use  `firewall_policy.0.stateful_rule_group_reference.0.override.0.action` because override is a struct (in the provider defined as a `list` with one element `action`), hence we still have to index it with `.0` to extract `action`. Only if the argument is of `TypeMap` like Tags can we directly do something like `override.action`.

Changes look good to me! I realised for `action` that it's actually `optional` and added a small suggestion to change that. Finally, we will need a changelog entry to record your changes. In this case support for a new argument. Example: https://github.com/hashicorp/terraform-provider-aws/pull/22396/files#diff-49cb1cdcd29781976b8252c4783ae8382940622737037d741804bf1b68936599

**Comment by teddylear** [comment]
@GlennChia Added changes. Thanks again for another quick review, please let me know if there's anything else I should change!

**Review by GlennChia — APPROVED** [review]
Changes LGTM! Approving this PR, pending Hashi Maintainer final review.

**Comment by teddylear** [comment]
@GlennChia Only requested re-approval so that hashi maintainer can see that this was reviewed already.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccNetworkFirewallFirewallPolicy_' PKG=networkfirewall ACCTEST_PARALLELISM=3
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/networkfirewall/... -v -count 1 -parallel 3  -run=TestAccNetworkFirewallFirewallPolicy_ -timeout 180m
=== RUN   TestAccNetworkFirewallFirewallPolicy_basic
=== PAUSE TestAccNetworkFirewallFirewallPolicy_basic
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulDefaultActions
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulDefaultActions
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulEngineOption
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulEngineOption
=== RUN   TestAccNetworkFirewallFirewallPolicy_updateStatefulEngineOption
=== PAUSE TestAccNetworkFirewallFirewallPolicy_updateStatefulEngineOption
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged
=== RUN   TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_multipleStatefulRuleGroupReferences
=== PAUSE TestAccNetworkFirewallFirewallPolicy_multipleStatefulRuleGroupReferences
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupPriorityReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupPriorityReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_statelessRuleGroupReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statelessRuleGroupReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_updateStatelessRuleGroupReference
=== PAUSE TestAccNetworkFirewallFirewallPolicy_updateStatelessRuleGroupReference
=== RUN   TestAccNetworkFirewallFirewallPolicy_multipleStatelessRuleGroupReferences
=== PAUSE TestAccNetworkFirewallFirewallPolicy_multipleStatelessRuleGroupReferences
=== RUN   TestAccNetworkFirewallFirewallPolicy_statelessCustomAction
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statelessCustomAction
=== RUN   TestAccNetworkFirewallFirewallPolicy_updateStatelessCustomAction
=== PAUSE TestAccNetworkFirewallFirewallPolicy_updateStatelessCustomAction
=== RUN   TestAccNetworkFirewallFirewallPolicy_multipleStatelessCustomActions
=== PAUSE TestAccNetworkFirewallFirewallPolicy_multipleStatelessCustomActions
=== RUN   TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction
=== PAUSE TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction
=== RUN   TestAccNetworkFirewallFirewallPolicy_tags
=== PAUSE TestAccNetworkFirewallFirewallPolicy_tags
=== RUN   TestAccNetworkFirewallFirewallPolicy_disappears
=== PAUSE TestAccNetworkFirewallFirewallPolicy_disappears
=== CONT  TestAccNetworkFirewallFirewallPolicy_basic
=== CONT  TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupPriorityReference
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged
--- PASS: TestAccNetworkFirewallFirewallPolicy_basic (155.43s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceManaged (157.08s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupPriorityReference (184.84s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_multipleStatefulRuleGroupReferences
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupOverrideActionReference (147.50s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupPriorityReference (170.73s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_updateStatelessCustomAction
--- PASS: TestAccNetworkFirewallFirewallPolicy_multipleStatefulRuleGroupReferences (172.47s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_disappears
--- PASS: TestAccNetworkFirewallFirewallPolicy_disappears (137.79s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_tags
--- PASS: TestAccNetworkFirewallFirewallPolicy_updateStatefulRuleGroupReference (199.76s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction
--- PASS: TestAccNetworkFirewallFirewallPolicy_tags (147.45s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_multipleStatelessCustomActions
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReferenceAndCustomAction (276.37s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_multipleStatelessRuleGroupReferences
--- PASS: TestAccNetworkFirewallFirewallPolicy_updateStatelessCustomAction (564.90s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statelessCustomAction
--- PASS: TestAccNetworkFirewallFirewallPolicy_multipleStatelessCustomActions (282.56s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_updateStatefulEngineOption
--- PASS: TestAccNetworkFirewallFirewallPolicy_multipleStatelessRuleGroupReferences (205.58s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_statelessCustomAction (153.40s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulEngineOption
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulRuleGroupReference (170.02s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_updateStatelessRuleGroupReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulEngineOption (167.31s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statefulDefaultActions
--- PASS: TestAccNetworkFirewallFirewallPolicy_updateStatefulEngineOption (302.76s)
=== CONT  TestAccNetworkFirewallFirewallPolicy_statelessRuleGroupReference
--- PASS: TestAccNetworkFirewallFirewallPolicy_updateStatelessRuleGroupReference (181.77s)
--- PASS: TestAccNetworkFirewallFirewallPolicy_statefulDefaultActions (163.15s)
--- PASS: TestAccNetworkFirewallFirewallPolicy_statelessRuleGroupReference (166.69s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/networkfirewall	1398.541s
```

**Comment by ewbankkit** [comment]
@teddylear Thanks for the contribution :tada: :clap:.

**Pull request merged**
